pyfr/backends/opencl/blasext.py

```diff
@@ -50,33 +50,43 @@ def errest(self, x, y, z, *, norm):
         if x.traits != y.traits != z.traits:
             raise ValueError('Incompatible matrix types')
 
-        cnt = x.leaddim*x.nrow
-        dtype = x.dtype
-
-        # Norm type
-        reduce_expr = 'a + b' if norm == 'l2' else 'max(a, b)'
+        nrow, ldim, dtype = x.traits
+        ncola, ncolb = x.ioshape[1:]
+
+        # Determine the "blocksize"
+        block = (128, 1, 1)
+        # Determine the "gridsize"
+        grid = (int((ncolb + block[0] - 1)/block[0]), 1, 1)
+        # Empty result buffer on host, shape = (nvars, nblocks)
+        err_host = np.empty((ncola, grid[0]), dtype)
+        # Device memory allocation
+        err_dev = cl.Buffer(self.backend.ctx, cl.mem_flags.READ_WRITE,
+                            err_host.nbytes)
+
+        # Get the kernel template
+        src = self.backend.lookup.get_template('errest').render(
+            norm=norm, ncola=ncola, sharesz=block[0]
+        )
 
         # Build the reduction kernel
-        rkern = ReductionKernel(
-            self.backend.ctx, dtype, neutral='0', reduce_expr=reduce_expr,
-            map_expr='pow(x[i]/(atol + rtol*max(fabs(y[i]), fabs(z[i]))), 2)',
-            arguments='__global {0}* x, __global {0}* y, __global {0}* z, '
-                      '{0} atol, {0} rtol'.format(npdtype_to_ctype(dtype))
-        )
+        rkern = self._build_kernel(
+            'errest', src, [np.int32]*3 + [np.intp]*4 + [dtype]*2)
 
         class ErrestKernel(ComputeKernel):
             @property
             def retval(self):
-                return self._retarr.get()
+                cl.enqueue_copy(self.queue.cl_queue_comp, err_host, err_dev)
 
-            def run(self, queue, atol, rtol):
-                qcomp = queue.cl_queue_comp
-
-                xarr = Array(qcomp, cnt, dtype, data=x.data)
-                yarr = Array(qcomp, cnt, dtype, data=y.data)
-                zarr = Array(qcomp, cnt, dtype, data=z.data)
+                # Reduction across blocks outsourced to numpy
+                if norm == 'uniform':
+                    return np.max(err_host, axis=1)
```

Comment:
Outside the class define something line

  reducer = np.max if norm == 'uniform' else np.sum

then just do return reducer(err_host, axis=1).